### DISCUSSION TITLE
Async createEffect Workaround valid?

### DISCUSSION THREAD

**Discussion by floric** [solidjs/solid#1691] [Q&A] [answered]
Hey :)

I'm currently struggling with asynchronous reactions. Until now I used `createResource` everywhere where I needed properties with asynchronous function calls. But I have a different usecase now where I was thinking about the following pattern:

```
const [someProp, setSomeProp] = createSignal("abc");
const [asyncProp, setAsyncProp] = createSignal("xyz");

createEffect(() => {
  const updatedProp = someProp();
  const updatedOtherProp = otherProp();

  // define async function and bind with static signal values
  const updateAsyncProp = async () => {
    const result = await doAsyncCall(updatedProp, updatedOtherProp);
    setAsyncProp(result);
  };

  // call reactive function
  updateAsyncProp();
});
```

Does this also break reactivity because of the asynchronous context where I call `setAsyncProp`, or would this be okay as the caller? (I'm planning this pattern for client side only reactivity)

**Comment by atk** [comment, marked as answer]
The setter needs not be inside a reactive scope, as it is the getter of a signal that registers the reactive tracking, so this should work.

But what is your use case that you want to use this reduced re-implementation of createResource instead of using it directly?

**Reply by floric** [reply]
Thanks for your answer. :)  `createResource` often feels kind of strange as it seems to be highly built to fetch resources from some remote based on an input parameter and always return exactly on property. In my case I often prefer to trigger multiple updates based on some asynchronous call.

For example I also have this code right now in a different feature:

```
createResource(async () => {
    setAbc("123");
    setXyz(true);
    await myFunction(myProperty());
    setAbc("abc");
    setXyz(false);
    return null;
});
```
To be honest, you are right. This looks simpler than the following:
```
createEffect(() => {
    const updatedProp = myProperty();
    const updateAsyncProps = async () => {
        setAbc("123");
        setXyz(true);
        await myFunction(updatedProp);
        setAbc("abc");
        setXyz(false);
    };
    updateAsyncProps();
});
```
or:
```
createEffect(() => {
    const updatedProp = myProperty();
    new Promise(() => {
        setAbc("123");
        setXyz(true);
        await myFunction(updatedProp);
        setAbc("abc");
        setXyz(false);
    });
});
```
Maybe I'm just not used to use `createResource` as it feels somehow different to my intention to use `createEffect` just asynchronously. Somehow I would prefer to use the latter snippets.

**Comment by atk** [comment]
It looks like you're not even using the output of the promise, just the timing. In that case, you do not need the resource.

**Reply by floric** [reply]
Yes exactly. That's why I asked the question :) I'm aware that `createResource` doesn't really fit. That's why I wondered if using `createEffect` this way would be a valid solution. Maybe we could consider to provide this pattern as a generic `createEffectAsync`?

**Reply by atk** [reply]
Maybe this works for you: [`createAsyncMemo`](https://solid-primitives.netlify.app/package/memo#createasyncmemo).

**Reply by floric** [reply]
This is also build to return a value (and memoize it). But in the end a new potencial helper could look pretty similar.

**Reply by atk** [reply]
Instead of

```ts
const [abc, setAbc] = createSignal();
const [xyz, setXyz] = createSignal();
createEffect(() => {
    const updatedProp = myProperty();
    new Promise(() => {
        setAbc("123");
        setXyz(true);
        await myFunction(updatedProp);
        setAbc("abc");
        setXyz(false);
    });
});
```

you can use

```ts
const asyncProps = createAsyncMemo(async () => {
  await myFunction(myProperty());
  return { abc: "abc", xyz: false };
}, { abc: "123", xyz: false });
```
